From bee4ca89897766c4b68856eafe14f1a3dad34251 Mon Sep 17 00:00:00 2001
From: Tomas Kyjovsky <tkyjovsk@redhat.com>
Date: Thu, 25 Jun 2020 11:31:05 +0200
Subject: [PATCH] KEYCLOAK-14872 CL DoS - read-timetout of the HTTP listener
 set to 30000 ms - read-timetout of the HTTPS listener set to 30000 ms -
 max-pool-size of the KeycloakDS datasource set to 100 connections

---
 .../content/bin/migrate-domain-clustered.cli  | 20 +++++++++++++++++++
 .../content/bin/migrate-domain-standalone.cli | 20 +++++++++++++++++++
 .../content/bin/migrate-standalone-ha.cli     | 20 +++++++++++++++++++
 .../content/bin/migrate-standalone.cli        | 20 +++++++++++++++++++
 .../keycloak-datasources.xml                  |  3 +++
 .../subsystem-templates/keycloak-undertow.xml |  4 ++--
 6 files changed, 85 insertions(+), 2 deletions(-)

diff --git a/distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-domain-clustered.cli b/distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-domain-clustered.cli
index e5c4f2b8df48..6765dfa987ab 100644
--- a/distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-domain-clustered.cli
+++ b/distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-domain-clustered.cli
@@ -695,6 +695,26 @@ if (result != fixed) of /profile=$clusteredProfile/subsystem=keycloak-server/spi
     end-try
 end-if
 
+# Migrate from 8.0.0 to 9.0.5
+
+if (result == 120000 || result == undefined) of /profile=$clusteredProfile/subsystem=undertow/server=default-server/http-listener=default/:read-attribute(name=read-timeout)
+    echo Updating value of http listener read-timeout.
+    /profile=$clusteredProfile/subsystem=undertow/server=default-server/http-listener=default/:write-attribute(name=read-timeout,value=30000)
+    echo
+end-if
+
+if (result == 120000 || result == undefined) of /profile=$clusteredProfile/subsystem=undertow/server=default-server/https-listener=https/:read-attribute(name=read-timeout)
+    echo Updating value of https listener read-timeout.
+    /profile=$clusteredProfile/subsystem=undertow/server=default-server/https-listener=https/:write-attribute(name=read-timeout,value=30000)
+    echo
+end-if
+
+if (result == 20 || result == undefined) of /profile=$clusteredProfile/subsystem=datasources/data-source=KeycloakDS/:read-attribute(name=max-pool-size)
+    echo Updating value of datasource max-pool-size.
+    /profile=$clusteredProfile/subsystem=datasources/data-source=KeycloakDS/:write-attribute(name=max-pool-size,value=100)
+    echo
+end-if
+
 # Migrate from 10.0.2 to 11.0.0 (migration changes for infinispan update from 9.4.18.Final to 10.1.8.Final)
 
 if (result != org.keycloak.keycloak-model-infinispan) of /profile=$clusteredProfile/subsystem=infinispan/cache-container=keycloak:read-attribute(name=module)
diff --git a/distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-domain-standalone.cli b/distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-domain-standalone.cli
index 652bbb81804c..f5287998203a 100644
--- a/distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-domain-standalone.cli
+++ b/distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-domain-standalone.cli
@@ -596,6 +596,26 @@ if (result != fixed) of /profile=$standaloneProfile/subsystem=keycloak-server/sp
     end-try
 end-if
 
+# Migrate from 8.0.0 to 9.0.5
+
+if (result == 120000 || result == undefined) of /profile=$standaloneProfile/subsystem=undertow/server=default-server/http-listener=default/:read-attribute(name=read-timeout)
+    echo Updating value of http listener read-timeout.
+    /profile=$standaloneProfile/subsystem=undertow/server=default-server/http-listener=default/:write-attribute(name=read-timeout,value=30000)
+    echo
+end-if
+
+if (result == 120000 || result == undefined) of /profile=$standaloneProfile/subsystem=undertow/server=default-server/https-listener=https/:read-attribute(name=read-timeout)
+    echo Updating value of https listener read-timeout.
+    /profile=$standaloneProfile/subsystem=undertow/server=default-server/https-listener=https/:write-attribute(name=read-timeout,value=30000)
+    echo
+end-if
+
+if (result == 20 || result == undefined) of /profile=$standaloneProfile/subsystem=datasources/data-source=KeycloakDS/:read-attribute(name=max-pool-size)
+    echo Updating value of datasource max-pool-size.
+    /profile=$standaloneProfile/subsystem=datasources/data-source=KeycloakDS/:write-attribute(name=max-pool-size,value=100)
+    echo
+end-if
+
 # Migrate from 10.0.2 to 11.0.0 (migration changes for infinispan update from 9.4.18.Final to 10.1.8.Final)
 
 if (result != org.keycloak.keycloak-model-infinispan) of /profile=$standaloneProfile/subsystem=infinispan/cache-container=keycloak:read-attribute(name=module)
diff --git a/distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-standalone-ha.cli b/distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-standalone-ha.cli
index 92bcdb612d8a..8baf80bfa03c 100644
--- a/distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-standalone-ha.cli
+++ b/distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-standalone-ha.cli
@@ -774,6 +774,26 @@ if (result != fixed) of /subsystem=keycloak-server/spi=hostname/:read-attribute(
     end-try
 end-if
 
+# Migrate from 8.0.0 to 9.0.5
+
+if (result == 120000 || result == undefined) of /subsystem=undertow/server=default-server/http-listener=default/:read-attribute(name=read-timeout)
+    echo Updating value of http listener read-timeout.
+    /subsystem=undertow/server=default-server/http-listener=default/:write-attribute(name=read-timeout,value=30000)
+    echo
+end-if
+
+if (result == 120000 || result == undefined) of /subsystem=undertow/server=default-server/https-listener=https/:read-attribute(name=read-timeout)
+    echo Updating value of https listener read-timeout.
+    /subsystem=undertow/server=default-server/https-listener=https/:write-attribute(name=read-timeout,value=30000)
+    echo
+end-if
+
+if (result == 20 || result == undefined) of /subsystem=datasources/data-source=KeycloakDS/:read-attribute(name=max-pool-size)
+    echo Updating value of datasource max-pool-size.
+    /subsystem=datasources/data-source=KeycloakDS/:write-attribute(name=max-pool-size,value=100)
+    echo
+end-if
+
 # Migrate from 10.0.2 to 11.0.0 (migration changes for infinispan update from 9.4.18.Final to 10.1.8.Final)
 
 if (result != org.keycloak.keycloak-model-infinispan) of /subsystem=infinispan/cache-container=keycloak:read-attribute(name=module)
diff --git a/distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-standalone.cli b/distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-standalone.cli
index e82e198c4536..5ff3b048ab3c 100644
--- a/distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-standalone.cli
+++ b/distribution/feature-packs/server-feature-pack/src/main/resources/content/bin/migrate-standalone.cli
@@ -639,6 +639,26 @@ if (result != fixed) of /subsystem=keycloak-server/spi=hostname/:read-attribute(
     end-try
 end-if
 
+# Migrate from 8.0.0 to 9.0.5
+
+if (result == 120000 || result == undefined) of /subsystem=undertow/server=default-server/http-listener=default/:read-attribute(name=read-timeout)
+    echo Updating value of http listener read-timeout.
+    /subsystem=undertow/server=default-server/http-listener=default/:write-attribute(name=read-timeout,value=30000)
+    echo
+end-if
+
+if (result == 120000 || result == undefined) of /subsystem=undertow/server=default-server/https-listener=https/:read-attribute(name=read-timeout)
+    echo Updating value of https listener read-timeout.
+    /subsystem=undertow/server=default-server/https-listener=https/:write-attribute(name=read-timeout,value=30000)
+    echo
+end-if
+
+if (result == 20 || result == undefined) of /subsystem=datasources/data-source=KeycloakDS/:read-attribute(name=max-pool-size)
+    echo Updating value of datasource max-pool-size.
+    /subsystem=datasources/data-source=KeycloakDS/:write-attribute(name=max-pool-size,value=100)
+    echo
+end-if
+
 # Migrate from 10.0.2 to 11.0.0 (migration changes for infinispan update from 9.4.18.Final to 10.1.8.Final)
 
 if (result != org.keycloak.keycloak-model-infinispan) of /subsystem=infinispan/cache-container=keycloak:read-attribute(name=module)
diff --git a/wildfly/server-subsystem/src/main/resources/subsystem-templates/keycloak-datasources.xml b/wildfly/server-subsystem/src/main/resources/subsystem-templates/keycloak-datasources.xml
index 5fb32ea190aa..f7b199c983d7 100755
--- a/wildfly/server-subsystem/src/main/resources/subsystem-templates/keycloak-datasources.xml
+++ b/wildfly/server-subsystem/src/main/resources/subsystem-templates/keycloak-datasources.xml
@@ -36,6 +36,9 @@
                     <user-name>sa</user-name>
                     <password>sa</password>
                 </security>
+                <pool>
+                    <max-pool-size>100</max-pool-size>
+                </pool>
             </datasource>
             <drivers>
                 <driver name="h2" module="com.h2database.h2">
diff --git a/wildfly/server-subsystem/src/main/resources/subsystem-templates/keycloak-undertow.xml b/wildfly/server-subsystem/src/main/resources/subsystem-templates/keycloak-undertow.xml
index faebb13f4ee6..4ef3bcc67755 100644
--- a/wildfly/server-subsystem/src/main/resources/subsystem-templates/keycloak-undertow.xml
+++ b/wildfly/server-subsystem/src/main/resources/subsystem-templates/keycloak-undertow.xml
@@ -28,8 +28,8 @@
         <buffer-cache name="default"/>
         <server name="default-server">
             <?AJP?>
-            <http-listener name="default" socket-binding="http" redirect-socket="https" enable-http2="true"/>
-            <https-listener name="https" socket-binding="https" security-realm="ApplicationRealm" enable-http2="true"/>
+            <http-listener name="default" socket-binding="http" redirect-socket="https" enable-http2="true" read-timeout="30000"/>
+            <https-listener name="https" socket-binding="https" security-realm="ApplicationRealm" enable-http2="true" read-timeout="30000"/>
             <host name="default-host" alias="localhost">
                 <location name="/" handler="welcome-content"/>
                 <http-invoker security-realm="ApplicationRealm"/>
